### ISSUE TITLE
Allow specific version of container run time in minikube

### ISSUE THREAD

**Issue by medyagh** [kubernetes/minikube#6432]
it would be nice that we allow minikube users create a cluster with exact specific version of container run time they desire.

such as Docker version 18 ...19  or container run time.
this would be a good opportunity to decouple installation logic of container run time tools from minikube. and maybe use a light config management system !

this would mean minikube would be shipped with a default tested version but also allow users to specify their own desired version

**Comment by afbjorklund** [comment]
> this would be a good opportunity to decouple installation logic of container run time tools from minikube. and maybe use a light config management system !

With the current minikube distribution and iso, there is not a good way of accomplishing this...
The buildroot packaging system is based on Makefiles, so there is no way of installing packages like there was on the Boot2Docker ISO. For docker-machine the goal was accomplished by using different images (one for each version), but that is not really feasible unless we start to make one iso for each runtime and version... Even for the supported alternative operating systems (ubuntu and centos - see #4733), it is not always easy to even installing **one** single runtime version - let alone "all" of them.

We could accomplish this by moving away from libmachine and instead provision naked machines, and then install the container runtime ourselves - similar to how we install kubeadm and friends today.

**Comment by afbjorklund** [comment]
Related to [roadmap](https://minikube.sigs.k8s.io/docs/contributing/roadmap/) items:
- _Generic (SSH) driver support_
- _Prototype post-libmachine implementation of minikube_

It would be good to get some statistics on which container runtimes and versions that are being used (or requested), previous reports say at least 80% of kubernetes users are running with the default `docker`.

* https://sysdig.com/blog/sysdig-docker-usage-report-2017/
  43% Kubernetes, of which 100% Docker 
* https://sysdig.com/blog/2018-docker-usage-report/
  51% Kubernetes, of which 83% Docker
* https://sysdig.com/blog/sysdig-2019-container-usage-report/
  89% Kubernetes, of which 79% Docker

**Comment by medyagh** [comment]
I think this would be do-able once we unify VM and container image to avoid doing it two times

**Comment by medyagh** [comment]
it would still be nice to have this feature eventually

**Comment by medyagh** [comment]
I think this could be done once we consolidated our VM/Docker Image/ISO so we could use same mechanisim to install specific version of each tool

**Comment by afbjorklund** [comment]
Currently this is not possible for podman/cri-o (without our own binaries), since they don't support / keep older versions in their package repositories

**Comment by afbjorklund** [comment]
I don't think it actually works with the Ubuntu packages either (not guaranteed to keep older versions), so would have to build from upstream sources

**Comment by afbjorklund** [comment]
We are now using the docker.com packages for docker, and the cri-o.io (kubic.opensuse.org) packages for podman.

But we haven't pinned them down, only podman was pinned to < 2 (interpreted as 1.9.3) due to API incompatibilies.

Eventually it will use latest and greatest, currently:

`docker-ce | 5:19.03.13~3-0~ubuntu-focal | https://download.docker.com/linux/ubuntu focal/stable amd64 Packages`

`containerd.io |    1.3.7-1 | https://download.docker.com/linux/ubuntu focal/stable amd64 Packages`

`podman |    2.1.1~2 | https://download.opensuse.org/repositories/devel:/kubic:/libcontainers:/stable/xUbuntu_20.04  Packages`

`cri-o-1.17 |   1.17.4~1 | https://download.opensuse.org/repositories/devel:/kubic:/libcontainers:/stable/xUbuntu_20.04  Packages`

Unless we need to pin down the major version again <sigh>

* Docker 20.10.0 Beta

* Containerd 1.4.1

* Podman 2.2.0 RC

* CRI-O 1.19.0

Chances are quite high that we **don't** want users picking any random version, since it lessens the chance of working.

So we will most likely go back to picking a version of each, and that will proably not be the "nightly" but a LTS version...

* Docker 19.03.x
* Containerd 1.3.x
* Podman 2.1.*
* CRI-O 1.18.x

The problem here being that there _are_ no long term support editions for Docker Community Edition or for Podman.

Once the new version is out (with breaking changes and bugs), that usually means end-of-life for the prior versions.

**Comment by afbjorklund** [comment]
https://sysdig.com/blog/sysdig-2021-container-security-usage-report/
91% Kubernetes, 50% Docker

**Comment by afbjorklund** [comment]
https://sysdig.com/blog/2022-cloud-native-security-usage-report/
96% Kubernetes, 46% Docker

**Comment by nirs** [comment]
Not clear who need this feature and we don't have capacity.
